README.md
# vectogebra v 0.0.10 - 29 May 2022

### Python module for vector algebra

easy to use vector algebra library for python, that lets ypu work with vectors in an efficient way.
apart from core vector object, many other vector operations are supported.
these can be imported from vectogebra.utilities.

this library was made by keeping its applications in Physics in mind (Mechanics, Optics, etc.)

- does not depend on any external libraries except math library.
- fully functional
- easy to use
- supports nearly all vector operations
- beginner friendly
- physics friendly
- Open for modifications

## 🌠 Install

```bash
pip install --upgrade vectogebra
```

### ⭐ Start by importing the vector class

```python
import vectogebra.vector as vect
```

_OR_

```python
from vectogebra import vector as vect
```

### ⭐ Also import useful utility functions

```python
import vectogebra.utitlies as vut
```

_OR_

```python
from vectogebra import utilities as vut
```

### ⭐ import functions for geometry applications :

```python
from vectogebra import geometry as geo
```

_OR_

```python
import vectogebra.geometry as geo
```

## 🧾 Description of the module

this module currently have two components : one is `vectogebra.vector`, which is the vector class (boject) defination. it contains the basic functionality.
the second component, `vectogebra.utilities` contains useful functions that are defined for the above mentioned vector class like, function to find angle between
two vectors, etc.

Create a vector object :

```python
import vectogebra.vector as vect

v1 = vect(1,2,3)
```

### 🔢 Algebric operations :

#### 1. Addition

consider two(or more) vectors : a,b,...
their sum will be given by :
`s = a + b + ...`
sum `s` will also be a vector object.

#### 2. Subtraction

Vectors can be subtracted using the minus (`-`) operator.

example :

`s = a - b + c - d + ...`

resultant `s` will also be a vector object.

#### 3. Dot product / scalar product and scalar multiplication

the `*` operator will be used for dot product, or multiplication by a scalar.

example :

`p = a * b * c * d * ...` is same as "a dot b dot c dot ...".

`p = 5*v` OR `v*5` is same as "scalar 5 multiplied to vector v".

#### 4. Cross product / vector multiplication

the `^` operator (XOR operator) will be used for cross product, or vector product.

example :

`p = a^b` is same as "p equals a cross b".

#### 5. division by a scalar

simply divide a vector by a scalar.
NOTE : division by zero or division vector is not supported.

example :

`p = v / 5` is same as "p equals v divided by 5".

### ❌✔️ Logical operations :

#### 1. Equality

`a == b` returnes True when a and b are equal in magnitude and direction. else, it returns False

#### 2. Inequality

`a != b` have its usual meaning

#### 3. grater / lesser

the **magnitude** of the vectors can be compared using common logical operators.

```python
# a and b are vectors
a > b
a < b
a >= b
a <= b
```

## Attributes of the vector object

### Components

for a vector v1,

1. `v1.x` **OR** `v1.i`
2. `v1.y` **OR** `v1.j`
3. `v1.z` **OR** `v1.k`

### Magnitude

4. `v1.magnitude` **OR** `v1.mod`

### Magnitude squared (useful when precesion is required)

5.  `v1.magnitude_squared` **OR** `v1.mod_squared`

### Type

6. `v1.type` different from `type(v1)`

## 🚀 Vectogebra's Utitlies (vut)

important utility functions for the vector object.
**import** :

```python
import vectogebra.utilities as vut
```

| S. no. | function                                       | Return value                                                  |
| ------ | ---------------------------------------------- | ------------------------------------------------------------- |
| 1.     | `vut.angle(v1,v2)`                             | angle between v1 and v                                        |
| 2.     | `vut.dot(v1,v2)`                               | dot product (or scalar product) of vectors v1 and v2          |
| 3.     | `vut.cross(v1,v2)`                             | cross product (or vector product) of v1 and v2                |
| 4.     | `vut.magnitude(v1)`                            | magnitude of v1 and v2                                        |
| 5.     | `vut.is_perpendicular(v1,v2)`                  | True when v1 is **perpendicular** to v2 else it returns False |
| 6.     | `vut.is_parallel(v1,v2)`                       | True whe v1 is parallel to ve else False                      |
| 7.     | `vut.scalar_component_parallel(v1,v2)`         | _Magnitude_ of **component** of v1 **parallel** to v2         |
| 8.     | `vut.scalar_component_perpendicular(v1,v2)`    | _Magnitude_ of **component** of v1 **perpendicular** to v2    |
| 9.     | `vut.vector_component_parallel(v1,v2)`         | _Vector_ **component** of v1 **parallel** to v2               |
| 10.    | `vut.vector_component_perpendicular(v1,v2)`    | _Vector_ compoment of v1 perpendicular to v2                  |
| 11.    | `vut.unit_vector(v)` **OR** `vut.direction(v)` | Returns the **unit vector** parallel to v                     |
| 12.    | `vut.dot(v1,v2)`                               | **dot** product                                               |
| 13.    | `vut.cross(v1,v2)`                             | **cross** product                                             |
| 14.    | `vut.parallelogram_area(v1,v2)`                | parallelogram's area formed by joining v1 and v2 tail to tail |
| 15.    | `vut.box(a,b,c)`                               | Box product or scalar triple product                          |
| 16.    | `vut.collinear(a,b,c)`                         | returns True if a,b,c are collinear                           |
| ⭐     | **Conversions**                                |                                                               |
| 17.    | `vut.vector_to_list(v)`                        | a list of the components of v                                 |
| 18.    | `vut.vector_to_dict(v)`                        | a dictionary of the components of v                           |
| 19.    | `vut.vector_to_tuple(v)`                       | a tuple of the components of v                                |
| 20.    | `vut.list_to_vector(l)`                        | a vector object from a list of components                     |
| 21.    | `vut.dict_to_vector(d)`                        | a vector object from a dictionary of components               |
| 22.    | `vut.tuple_to_vector(t)`                       | a vector object from a tuple of components                    |
| 23.    | `vut.proportional(v1, v2)`                     | `True` if two vectors are proportional, otherwise : `False`   |

(more to come)

## 📈 Geometry related functions :

consider vectors `a`, `b` and `c` :

| S. no. | function                 | Return value                                                                                            | ⚠ Warning                                                      | 📃 Special instructions                                                |
| ------ | ------------------------ | ------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------- | ---------------------------------------------------------------------- |
| 1.     | `divider(a,b,m,n)`       | a position vector of a point that divides the line segment joining `a` and `b` by in the ratio `m`:`n`. | atleast one of `m` or `n` must be non-zero                     |                                                                        |
| 2.     | `distance(a,b)`          | distance between `a` and `b`                                                                            |                                                                |                                                                        |
| 3.     | `area_line(a,b)`         | signed area under the line segment joining `a` and `b`.                                                 | x-y plane only (or area under projection of line on x-y plane) |                                                                        |
| 4.     | `area_polygon(*args)`    | Signed area of polygon whose vertices are given as input                                                | x-y plane only (or projection on x-y plane)                    | if vertices listed in cyclic manner the area will be +ve else -ve.     |
| 5.     | `area_triangle(a, b, c)` | Area of triangle with vertices `a` , `b`, and `c`                                                       |                                                                |                                                                        |
| 6.     | `coplanar(*args)`        | `True` : If all the points are coplanar else `False`                                                    | Arguments must be **position vectors.**                        | Keep in mind the difference between a _vector_ and a _position vector_ |

## Useful classes for 3-Dimentional Geometry

### line :

represents a line in a 3D space

#### construction of a line :

1. Two - point form

- two points (positions vectors) or list/tuple/dict/string-representation of coordinates must be given to the constructor
- examples :

  ```py
  import vectogebra.vector as vect
  import vectogebra.geometry as geo

  line1 = geo.line(vect(5,8,9),vect(1,6,3))
  line2 = geo.line((5,8,9),(1,6,3)) # can replace tuples whit lists.
  line3 = geo.line('5 8 9', '1 6 3')
  # dict of components can also be used.

  ```

2. Point-direction form :

- one point on the line, and the direction vector can be used to define a line :
  ```py
  line1 =  geo.line(p = (1,5,3), d= (5,2,8))
  # instead of thuples, list, or string 'x y z' or vectogebra's vector can also be used.
  ```

#### Attributes :

1. for direction : `d` or `dir` or `direction`
2. for point : `p` or `pt` or `point`

#### class line have many useful functions defined. check them in the source code. they will be listed here later.

### plane :

#### Methods of defining a plane

1. by a point and a normal vector :

```py
plane(point=(x,y,z),normal=(a,b,c))
```

- vector object or list of componens can also be used instead of tuples.
- instead of `point`, `p` or `pt` can also be used.
- instead of `normal`, `n` or `norm` can also be used.

2. by three points :

```py
plane((x1,y1,z1),(x2,y2,z2),(x3,y3,z3))
```

vector `vectogebra.vector(x,y,z)` object or lists of components `[x,y,z]` can also be used instead of tuples.

#### Attributes :

- position vector of the point on the plane : `p` or `pt` or `point`
- normal vector of the plane : `n` or `normal`

---

Author: Mohammad Maasir

License: MIT

date-created: 8th of May, 2022

PyPi :https://pypi.org/project/vectogebra/

---

_Copyright © 2022 Mohammad Maasir_


CHANGELOG.md
# CHANGELOG

record of major changes to the project including bug fixes will be stored here.

## [0.0.6]

### the library is now comlpetely functional.

- optimised the directory structure.
- importing vector class is now easy.
- resolved the issue of angle between parallel vectors.

## [0.0.7]

### updates in vector class

- 2D vector can now be defined by passing two arguments and third will automatically be set to 0.
- `vector()` can now accept single argument as a list / dict or tuple of x, y, and z components.
- ### new attributes:
  1. `vector.angle_x` - angle between vector and x axis.
  2. `vector.angle_y` - angle between vector and y axis.
  3. `vctor.angle_z` - angle between vector and z axis.
     (above angles will be in radians) suffix `_deg` at the end of the attribute to get angle in degrees.
- ### define vector using string !
  vector(`x y z`) will return vector object : vector(x, y, z)

### updates in vut

- polar vectors are now supported.
  to define a vector using polar argument, use vut.polar_to_vector(r, theta, atype)
  atype can be 'deg' or 'degree' or 'rad' or 'radian' default is 'rad'. example:
  ```python
    >>> vut.polar_to_vector(1, 90, 'deg')
    >>> vut.polar_to_vector(1, math.pi/2, 'rad')
  ```
- polar exportes added
  vector can be exported to a polar representation by using the following functions:
  ```python
     vut.polar_to_vector(vector, plane, atype)
  ```
  - plane can be 'xy' or 'yz' or 'zx'
  - atype can be 'deg' or 'degree' or 'rad' or 'radian' default is 'rad'.
  - default plane is 'xy', and default atype is 'deg' in this case.
  - angle will be measured from the +ve x- axis in case of xy and yz plane, and with +ve y- axis in the case of yz plane.

### new module for geometry

this module contains functions for coordinate geometry calculations using vectors (Position Vectors).
this module is an application of vector.py and utilities.py

```python
import vectogebra.geometry as geo
```

defined the following functions :-

1. `divider(a,b,m,n)` : returns a point that divides the line segment joining `a` and `b` by in the ratio `m`:`n`.
2. `distance(a,b)` : returns the distance between `a` and `b`
3. `area_line(a,b)` : signed area under the line segment joining a and b.
4. `area_polygon` : signed area of polygon with any number of vertices.
5. `area_triangle` : area formed by a triangle with vertices `a` , `b` and `c`.

## [0.0.8]

### 1. changes in vector class

- `round()` method defined for the vector class.
- Conversions to `list` , `dict`, and `tuple` using attribute names :
  _[26 May, 2022]_
  ```python
    v = vectorgbra.vector(1,2,3)
    v.list # returns list of components in order of x, y, z.
    v.dict # returns dict of components in order of x, y, z.
    v.tuple # returns tuple of components in order of x, y, z.
  ```

### 2. changes in geometry.py mudule:

- `coplanar(*args)` function defined. this function returns `True` if the given points are coplanar. else returns `False`. - _23 May, 2022_

  [*24 May 2022*]

- `area_polygon(*args)` and `area_line(a,b)` functions defined.

- `centroid(*args)` function defined which returns the centroid of the polygon.

- Class `line` defined. this class represents a line in 3-dimensional space. its methods include :-

  [ _25 May, 2022_ ]

  - `self.parallel(line)` : returns `True` if the given line is parallel to the current line.
  - `self.intersection(line)` : returns the intersection point of the current line and the given line.
  - `self.distance(other)` : returns the (shortest)distance between the current line and the given line or point.
  - `self.intersect(line)` returns `True` if given line intersects with the current line.
  - `self.includes` : returns `True` if the given point is on the current line.

  [ _26 May, 2022_ ]

  - To construct a line object instance :
    ```python
    # if p1 and p2 are points on the line
    line1 = vectogebra.geometry.line(p1, p2)
    # if a is a point on the line and b is the direction vector :
    line2 = vectogebra.geometry.line(point = a, direction = b)
    # in place of point, p or pt can also be used.
    # in place of direction, d or dir can also be used.
    ```
  - equality of line also defined : if two lines have same direction and one point in common, they are equal.
  - `line.__eq__(other)` : returns `True` if the given line is equal to the current line. else returns `False`.

    ```python
     >>>import vectogebra
     >>>line1 = vectgebra.geometry.line(p=vectogebra.vector(0,0,0),d=vectogebra.vector(1,1,1))
     >>>line2 = vectogebra.geometry.line(p=vectogebra.vector(5,5,5),d=vectogebra.vector(99,99,99))
     >>>line1 == line2
     >>>True
    ```

    [*27 May, 2022*]

  - defination of line can be done using tuples or strings also

    ```py
    import vectogebra.geometry as geo
    line1 = geo.line(p=(0,0,0),d=(1,1,1))
    # OR :
    line2 = geo.line(p='0 0 0',d='1 1 1')
    ```

  - Now line can also be constructed using `tuple` or `list` instead of `vector` :
    ```py
    line1 = vectogebra.geometry.line(p=(1,2,4),d=(15,12,33))
    # two point method also applicable :
    line2 = vectogebra.geometry.line((12,46,67),(12,98,65))
    # similarly, tuples can also be used.
    ```

- Class `plane` defined : Represents a plane in 3D space.

  [27 May, 2022]

  - different methods of defining the plane.
  - equality of planes.
  - intersection of plane with a line (result is a point.)
  - intersection of plane with a plane (result is a line)
  - function to check if a plane includes a given point or not.
  - function to check if a plane is parallel to a given plane or line.
  - distnce between a plane and a point or line or other plane.
  - now arguments need not be vectors, tuples/ lists / strings can also be used.

  ```py
  import vectogebra.geometry as geo
  plane1 = geo.plane(p=(0,0,0),n=(1,1,1))
  # OR
  plane2 = geo.plane(p='0 0 0',n='1 1 1')
  ```

### 3. changes in utilities.py mudule:


test/__init__.py
# this is an important file! even if it is empty.
# do NOT delete it.



test/test_algebra.py
"""
---------------------------------------------------------------------
vectogebra/test/test_algebra.py
---------------------------------------------------------------------
File under project "vectogebra"
---------------------------------------------------------------------
This file contains the test cases for the attributes of vector class.
---------------------------------------------------------------------
License: MIT License (see LICENSE in project's main directory)
copyright: (C) 2022 Mohammad Maasir
---------------------------------------------------------------------
- date created: 9th of May, 2022 (11:35 PM)
- last modified: 9th of May, 2022
---------------------------------------------------------------------
contributor(s): Mohammad Maasir
---------------------------------------------------------------------
github: github.com/maasir554/vectogebra/vectogrbra/vector.py
---------------------------------------------------------------------
report an issue at: github.com/maasir554/vectogebra/issues
---------------------------------------------------------------------
email: maasir554@gmail.com


"""

# NOTE: do NOT run this file directly.

# Instead, open console or terminal and change dirctory to root directory of this project : 
# $ cd vectogebra
# then, run the following command:
# python -m unittest test/test_algebra.py 
# OR
# python -m unittest test.test_algebra
# OR 
# python -m unittest discover 

## and you should see an OK message.
## if not, then you have some test case(s) that failed.

#####################################################################

############## the following block is VERY important ################

#####################################################################
#------------------------------------------------------(result of a lot of research, trial and error)
import pathlib
import sys
currentDIR = pathlib.Path(__file__).resolve().parent
rootDIR = currentDIR.parent
vectogebraMAIN = rootDIR / 'src' /'vectogebra'
sys.path.append(str(vectogebraMAIN))
sys.path.append(str(rootDIR))
# os module can also be used here. but pathlib is easier to work with
#------------------------------------------------------  (added the src/vectogevra to the path for the test file to work)
import unittest

# from vector import vector as vect

from src.vectogebra.vector import vector as vect
# if you wish to run directly, you can use the following:(disable the above line first)
# from vector import vector as vect
v1 = vect(1,2,3)
v2 = vect(4,5,6)
v3 = vect(2,2,2)


class TestStringMethods(unittest.TestCase):
    
    # ########################################## #

    # ######### ALGEBRIC OPERATORS ############# #

    # ########################################## #

    def test_addition_operator(self):
        self.assertEqual(v1+v2, vect(5,7,9))
        self.assertEqual(v1+v2+v3, vect(7,9,11))

    def test_subctraction_operator(self):
        self.assertEqual(v1-v2, vect(-3,-3,-3))
        self.assertEqual(v1-v2-v3, vect(-5,-5,-5))

    def test_dot_operator(self):
        self.assertEqual(v1*v2, 32)
        self.assertEqual(v1*v2*v3, 32*v3)

    def test_cross_operator(self):
        self.assertEqual(v1^v2, vect(-3,6,-3))
        self.assertEqual(v1^v2^v3, vect(18,0,-18))  #same as v1^(v2^v3)
    
    def test_scalar_multiplication(self):
        self.assertEqual(v1*2, vect(2,4,6))
        self.assertEqual(2*v1, vect(2,4,6))

    def test_scalar_division(self):
        self.assertEqual(v1/2, vect(0.5,1,1.5))

    def test_unary_operators(self):
        self.assertEqual(-v1, vect(-1,-2,-3))
        self.assertEqual(-v2,vect(-4,-5,-6))
        self.assertEqual(-v3,vect(-2,-2,-2))
        #for + uunary operator
        self.assertEqual(+v1, vect(1,2,3))
        self.assertEqual(+v2,vect(4,5,6))
        self.assertEqual(+vect(-1,5,-56),vect(-1,5,-56))
    
    def test_round_function(self):
        a = vect(1.29,2.36,3.45)
        b = vect(1.0453,2.0453,3.0453)
        self.assertEqual(round(a), vect(1,2,3))
        self.assertEqual(round(b), vect(1,2,3))
        self.assertEqual(round(a,1), vect(1.3,2.4,3.5))



if __name__ == '__main__':
    unittest.main()


src/vectogebra/__init__.py
"""
    file under vectogebra - python module for vector algebra
    
    vectogrbra/__init__.py
    -------------------------
    This module contains the 'vector' class.
    -------------------------------------------------
    copyright: (C) 2022 Mohammad Maasir
    license: MIT License (see LICENSE in project's main directory)
    -------------------------------------------------
    date created: 9th of May, 2022 (3:09 AM)
    last modified: 9th of May, 2022
    -------------------------------------------------
    contributor(s): Mohammad Maasir
    -------------------------------------------------
    github: github.com/maasir554

"""

# importing vector class from vector.py
# the dot is necessary to avoid name conflict. 
# from .vector import vector

# exception handling is best solution to any problem
# here in
try:
    from .vector import vector  # this import is for the build version
except :
    pass        # simply ignore the error and the module will be imported 
# pass will simply ignore the import command here and test files will import the module by
# the local command written in them


"""
    to use vector object class, in your file, write: import vectogerbra.vector as v
    then you can use the vector class in your file
    example: v1 = v(1,2,3)

    to import the vector class in your file, you can use the following:
    ## import vectogerbra.vector as v [or any other alias name you want]
    ## from vectogerbra import vector as v [or any other alias name you want]
"""

# importing utility functions from utilities.py

# from .utilities import *



src/vectogebra/utilities.py
"""
    Utitlity functions for vector(x,y,z)
    ----------------------------------------------------------

    file under vectogebra - python module for vector algebra

    ----------------------------------------------------------

    vectogebra/src/utilities.py

    -------------------------
    This module contains the utility operations for vasious calculation related to vectors.
    
    requires: vectogebra/src/vector.py to be imported
    
    depends on: 
    1. vectogebra/src/vector.py
    2. math module 
    -------------------------------------------------
    copyright: (C) 2022 Mohammad Maasir
    license: MIT License (see LICENSE in project's main directory)
    -------------------------------------------------
    date created: 9th of May, 2022 (3:09 AM)
    last modified: 9th of May, 2022
    -------------------------------------------------

    contributor(s): Mohammad Maasir

    -------------------------------------------------

    github: github.com/maasir554/vectogebra/vectogrbra/utilities.py

    ----------------------------------------------------------------

    report an issue : https://github.com/maasir554/vectogebra/issues
    
    ---------------------------------------------------------------------
    
    email: maasir554@gmail.com

    ---------------------------------------------------------------------

    

"""

# ------------------------------------------------------(result of a lot of research, trial and error)
# sys.path is where python search for modules.

import pathlib
import sys

currentDIR = pathlib.Path(__file__).resolve().parent
sys.path.append(str(currentDIR))
# ------------------------------------------------------  (added the src/vectogebra to the path for the test file to work)


# importlib is imported for making imports easier.


import importlib


# importing math for trigonometric and inverse trigonometric functions :-


import math


# importing vector class from vector.py


try:
    from .vector import vector as v
except:
   v = importlib.import_module('vector').vector
#    from vector import vector as v

# print(v)
# print(v(1,1,1)*v(2,2,2))

# NOTE: the type of vector will also be reffered as alias i.e. in this case, v only.
# example: a = v(1,2,3)
# print(type(a) == v) => This will return True. NOTE: do NOT use "v" or 'v' or vector.v ,
# just use alias as it is i.e.: v
# if we would have taken alias as vect then we would refer to it as vect only. and not "vect" or 'vect' or vector.vect
# if we did not had used alias, then we would refer to it as vector.vector
# Please don't get confused when working outside this module. there will be different conventions.

##########################################################

############# _____  ANGLE RELATED _____ ##################

##########################################################

# for angle between two vectors options : "radians"(default) or "degrees"
def angle(v1, v2, format="radians"):
    # python is bad at multipilcation of irrational numbers hence, we use magnitude_square()
    # instead of magnitude() (as magnitude_squared is defined separately, without use of square roots.)
    # else we  get errors like 5.99999999999999999 != 6 xD
    if (v1*v2)**2 != (v1.magnitude_squared)*(v2.magnitude_squared) and type(v1) == type(v2) == v:
        if format == "radians":
            return math.acos((v1 * v2) / (v1.magnitude * v2.magnitude))
        elif format == "degrees":
            return math.degrees(math.acos((v1 * v2) / (v1.magnitude * v2.magnitude)))
        else:
            raise ValueError("third argument must be either 'radians' or 'degrees'")
    elif (v1*v2)**2 == (v1.magnitude_squared)*(v2.magnitude_squared) and type(v1) == type(v2) == v:
        return 0.0
    else :
        raise TypeError("First two arguments must be vectors and optionally, third argument must be either 'radians' or 'degrees'")

# function to check if the tow vectors are perpendicular or not
def is_perpendicular(v1, v2):
    if angle(v1, v2, 'degrees') == 90.0:
        return True
    else:
        return False


# function to check if the tow vectors are parallel or not
def is_parallel(v1, v2):
    if angle(v1, v2, 'degrees') == 0.0:
        return True
    else:
        return False


# function to obtain unit vector from a given vector
def unit_vector(v):
    if v.magnitude != 0 :
        return (v / v.magnitude)
    else:
        return v # zero vector has arbitrary direction.

def direction(v):
    return unit_vector(v)


###########################################################

################_________ ALGEBRA _______###################

###########################################################

# addition of n vectors
def add(*args):
    v0 = v(0, 0, 0)
    for i in args:
        if type(i) == v:
            v0 = v0 + i
        elif type(i) != v:
            print(type(i))
            print(type(v))
            raise TypeError(type(v), type(i), "Arguments must be vectors")
        else:
            return 'Something went wrong :-('
    return (v0)


# function for dot product
def dot(v1, v2):
    return v1 * v2


# function for cross product
def cross(v1, v2):
    return v1 ^ v2


# magnitude of a vector or a number
def magnitude(v):
    if type(v) == v:
        return v.magnitude
    elif type(v) == int or type(v) == float:
        return (v ** 2) ** 0.5
    else:
        raise TypeError("Argument must be a vector or a number")

def proportional(a:v or tuple or list or str or dict,b:v or tuple or list or dict or str)->bool:
    if type(a) == v:
        v1 = a
    
    elif type(a) == tuple or type(a) == list or type(a) == str or type(a) == dict:
        v1 = v(a)
    else :
        raise TypeError("Problem in function proportional.")

    if type(b) == v:
        v2 = b
    elif type(b) == dict or type(b) == list or type(b) == tuple or type(b) == str:
        v2 = v(b)
    else :
        raise TypeError("Problem in function proportional.")

    return abs(v1 ^ v2) == 0


###########################################################

################_________ GEOMETRY _______##################

###########################################################

# function to find area of parallelogram formed by two vectors
def parallelogram_area(a, b):
    return a.magnitude * b.magnitude * math.sin(angle(a, b, 'radians'))


# function to find volume of parallelopiped formed by three vectors or box product
def box(a, b, c):
    return a * (b ^ c)


# function to check if three vectors are coplanar
def coplanar(a, b, c):
    if box(a, b, c) == 0:
        return (True)
    else:
        return (False)


# scalar component of vector a along the direction of vector b or opposite to direction of vector b(if the angle is
# obtuse)
def scalar_component_collinear(a, b):
    if type(a) == v and type(b) == v:
        return a * unit_vector(b)
    else:
        raise TypeError("Arguments must be vectors")


# vector component of vector a along the direction of vector b or opposite to direction of vector b(if the angle is
# obtuse)
def vector_component_collinear(a, b):
    return a * unit_vector(b) * unit_vector(b)


# scalar component perpendicular to direction of vector b
def scalar_component_perpendicular(a, b):
    return (a ^ b) / b.magnitude


# vector component perpendicular to direction of vector b
def vector_component_perpendicular(a, b):
    return scalar_component_perpendicular(a, b) * unit_vector(b ^ (a ^ b))  # Thumb rule logics***


###########################################################

#############________ EXPORT A VECTOR _______##############

###########################################################


# dictionary from a vector 
def vector_to_dict(v):
    return {'x': v.x, 'y': v.y, 'z': v.z}


# list from a vector (it will contain its coordinates/components)
def vector_to_list(v):
    return [v.x, v.y, v.z]  # returning a list of coordinates/components arranged in order of x,y,z


# tuple from a vector (it will contain its coordinates/components)
def vector_to_tuple(v):
    return v.x, v.y, v.z

def vector_to_polar(v, plane="xy", atype="deg"):
    if plane == "xy" and atype == "deg" or atype == "degree": 
        return (v.magnitude, v.angle_x_deg)
    elif plane == "xy" and atype == "rad" or atype == "radian":
        return (v.magnitude, v.angle_x)
    elif plane == "xz" and atype == 'rad'or atype == 'radian':
        return (v.magnitude, v.angle_degrees_x)
    elif plane == "xz" and atype == 'deg'or atype == 'degree':
        return (v.magnitude, v.angle_x_deg)
    elif plane == "yz" and atype == 'rad'or atype == 'radian':
        return (v.magnitude, v.angle_degrees_y)
    elif plane == "yz" and atype == 'deg'or atype == 'degree':
        return (v.magnitude, v.angle_y_deg)
    else:
        raise ValueError("plane must be either 'xy', 'xz' or 'yz'")

###########################################################

############# CREATION OF VECTOR  #########################

###########################################################

# vector from a dictionary
def dict_to_vector(d):
    if type(d) == dict:
        return v(d['x'], d['y'], d['z'])
    else:
        raise TypeError("Argument must be a dictionary")


# vector from a list


def list_to_vector(l):
    if type(l) == list:

        return (v(l[0], l[1], l[2]))

    else:

        raise TypeError("Argument must be a list")


# vector from a tuple

def tuple_to_vector(t):
    if type(t) == tuple:
        return v(t[0], t[1], t[2])
    else:
        raise TypeError("Argument must be a tuple")


# 2d vector from polar coordinates


def polar_to_vector(r, theta, atype="rad"):
    if atype == "rad" or atype == "radians" or atype == "radian"  :                    
        if round(math.cos(theta),10) != 0.0 and round(math.sin(theta),10) != 0.0:
            return v(r*math.cos(theta), r*math.sin(theta))
        elif round(math.cos(theta),10) ==0.0:
            return v(0, r*math.sin(theta))
        elif round(math.sin(theta),10) ==0.0:
            return v(r*math.cos(theta), 0)
        
    elif atype == "deg" or atype == "degrees" or atype == "degree"  :
        phi = math.radians(theta)
        if round(math.cos(phi),10) != 0.0 and round(math.sin(phi),10) != 0.0:
            return v(r*math.cos(phi), r*math.sin(phi))
        elif round(math.cos(phi),10) ==0.0:
            return v(0, r*math.sin(phi))
        elif round(math.sin(phi),10) ==0.0:
            return v(r*math.cos(phi), 0)
          
    else:
        raise ValueError("Third argument must be either 'radians' or 'radian' or 'rad' or 'degrees' or 'degree' or 'deg'")

